test/cpp/tensorexpr/test_boundsinference.cpp

```diff
@@ -725,5 +725,175 @@ void testGetPotentialHazardsLoopSplit() {
+  VarHandle i("i", kInt);
+  VarHandle j("j", kInt);
+  VarHandle m("m", kInt);
+  VarHandle n("n", kInt);
+  auto storeA1 = Store::make(a_buf, {i, j}, Mul::make(Mul::make(i, j), 500));
+  auto forJ = For::make(j, 0, 100, storeA1);
+  auto forI = For::make(i, 0, 20, forJ);
+  auto storeA2 =
+      Store::make(a_buf, {m + 1, n}, Add::make(m, Mul::make(n, 100)));
+  auto forN = For::make(n, 0, 50, storeA2);
+  auto forM = For::make(m, 0, 20, forN);
+  auto par = Block::make({forI, forM});
+
+  tensorexpr::analysis::MemDependencyChecker analyzer;
+  par->accept(&analyzer);
+  ASSERT_TRUE(hasPartialOverlap(analyzer, forI, forM));
+  ASSERT_TRUE(hasPartialOverlap(analyzer, forM, forI));
+  ASSERT_TRUE(hasPartialOverlap(analyzer, forJ, forN));
+  ASSERT_TRUE(hasPartialOverlap(analyzer, forN, forJ));
+  ASSERT_TRUE(hasPartialOverlap(analyzer, storeA1, storeA2));
+  ASSERT_TRUE(hasPartialOverlap(analyzer, storeA2, storeA1));
+  ASSERT_TRUE(hasPartialOverlap(analyzer, forJ, storeA2));
+  ASSERT_TRUE(hasPartialOverlap(analyzer, storeA1, forM));
+}
+
+TEST(BoundsInference, HasPartialOverlap2DBufferWithNoOverlap) {
+  KernelScope kernel_scope;
+
+  // Input IR:
+  //   for (int i = 0; i < 20; i++) {
+  //     for (int j = 0; j < 100; j++) {
+  //       A[i,j] = i * j * 500;
+  //     }
+  //   }
+  //   for (int m = 0; m < 20; m++) {
+  //     for (int n = 0; n < 50; n++) {
+  //       A[m+20,n+100] = m + n * 100;
+  //     }
+  //   }
+  BufHandle a_buf("A", {20, 100}, kInt);
+  BufHandle b_buf("B", {20, 50}, kInt);
+  VarHandle i("i", kInt);
+  VarHandle j("j", kInt);
+  VarHandle m("m", kInt);
+  VarHandle n("n", kInt);
+  auto storeA1 = Store::make(a_buf, {i, j}, Mul::make(Mul::make(i, j), 500));
+  auto forJ = For::make(j, 0, 100, storeA1);
+  auto forI = For::make(i, 0, 20, forJ);
+  auto storeA2 =
+      Store::make(a_buf, {m + 20, n + 100}, Add::make(m, Mul::make(n, 100)));
+  auto forN = For::make(n, 0, 50, storeA2);
+  auto forM = For::make(m, 0, 20, forN);
+  auto par = Block::make({forI, forM});
+
+  tensorexpr::analysis::MemDependencyChecker analyzer;
+  par->accept(&analyzer);
+  ASSERT_FALSE(hasPartialOverlap(analyzer, forI, forM));
+  ASSERT_FALSE(hasPartialOverlap(analyzer, forM, forI));
+  ASSERT_FALSE(hasPartialOverlap(analyzer, forJ, forN));
+  ASSERT_FALSE(hasPartialOverlap(analyzer, forN, forJ));
+  ASSERT_FALSE(hasPartialOverlap(analyzer, storeA1, storeA2));
+  ASSERT_FALSE(hasPartialOverlap(analyzer, storeA2, storeA1));
+  ASSERT_FALSE(hasPartialOverlap(analyzer, forJ, storeA2));
+  ASSERT_FALSE(hasPartialOverlap(analyzer, storeA1, forM));
+}
+
+TEST(BoundsInference, HasPartialOverlapDifferentBuffers) {
+  KernelScope kernel_scope;
+
+  // Input IR:
+  //   for (int i = 0; i < 20; i++) {
+  //     for (int j = 0; j < 100; j++) {
+  //       A[i,j] = i * j * 500;
+  //     }
+  //   }
+  //   for (int m = 0; m < 20; m++) {
+  //     for (int n = 0; n < 50; n++) {
+  //       B[m,n] = m + n * 100;
+  //     }
+  //   }
```

Comment:
We probably also need to add tests checking read-after-write dependencies. E.g.
```
for i in 0..100:
  A[i] = i
for j in 0..100:
  B[j] = A[99-j]
```
I.e. these two loops cannot be fused.

Reply:
Added tests for both RAW and WAR dependencies in `test_boundsinference` and `test_loopnest`.